connman/src/dns-systemd-resolved.c

```diff
@@ -434,6 +445,288 @@ int __connman_dnsproxy_set_mdns(int index, bool enabled)
 	return 0;
 }
 
+static void setlinkdefaultroute_append(DBusMessageIter *iter, void *user_data)
+{
+	struct resolved_dbus_data *data = user_data;
+	dbus_bool_t val;
+
+	val = data->enabled;
+
+	DBG("SetLinkDefaultRoute: %d/%d", data->index, val);
+
+	dbus_message_iter_append_basic(iter, DBUS_TYPE_INT32, &data->index);
+	dbus_message_iter_append_basic(iter, DBUS_TYPE_BOOLEAN, &val);
+}
+
+static int resolved_set_default_route(int index, bool enabled)
+{
+	struct resolved_dbus_data data = { .index = index, .enabled = enabled };
+
+	if (!resolved_proxy)
+		return -ENOENT;
+
+	if (index < 0)
+		return -ENODEV;
+
+	if (!g_dbus_proxy_method_call(resolved_proxy, "SetLinkDefaultRoute",
+			setlinkdefaultroute_append, NULL, &data, NULL))
+		return -EINVAL;
+
+	return 0;
+}
+
+static int resolved_flush_cache()
+{
+	DBG("");
+
+	if (!resolved_proxy)
+		return -ENOENT;
+
+	if (!g_dbus_proxy_method_call(resolved_proxy, "FlushCaches",
+			NULL, NULL, NULL, NULL))
+		return -EINVAL;
+
+	return 0;
+}
+
+static void setrevertlink_append(DBusMessageIter *iter, void *user_data)
+{
+	int index = GPOINTER_TO_INT(user_data);
+
+	DBG("RevertLink: %d", index);
+
+	dbus_message_iter_append_basic(iter, DBUS_TYPE_INT32, &index);
+}
+
+static int resolved_disable_interface(int index)
+{
+	if (!resolved_proxy)
+		return -ENOENT;
+
+	if (index < 0)
+		return -ENODEV;
+
+	if (!g_dbus_proxy_method_call(resolved_proxy, "RevertLink",
+			setrevertlink_append, NULL, GINT_TO_POINTER(index),
+			NULL))
+		return -EINVAL;
+
+	return 0;
+}
+
+
+struct interface_search_data {
+	bool enable;
+	bool forced;
+	int index4;
+	int index6;
+	int vpn_index;
+};
+
+static bool interface_index_match(int index,
+				struct interface_search_data *search_data)
+{
+	if (search_data->forced)
+		return true;
+
+	if (index == search_data->index4 || index == search_data->index6) {
+		DBG("DNS server of service %d", index);
+		return true;
+	}
+
+	if (index == search_data->vpn_index) {
+		DBG("DNS server of a VPN %d", index);
+		return true;
+	}
+
+	return false;
+}
+
+static void toggle_interface_dns(struct dns_interface *iface, bool enable)
+{
+	DBG("");
+
+	if (!iface || is_empty(iface))
+		return;
+
+	if (iface->enabled == enable)
+		return;
+
+	iface->enabled = enable;
+
+	if (!enable) {
+		DBG("disabling interface %d", iface->index);
+		resolved_disable_interface(iface->index);
+	} else {
+		DBG("re-enabling interface %d", iface->index);
+		set_systemd_resolved_values(iface);
+		if (is_empty(iface)) {
```

Comment:
This cannot happen, because of the earlier check. I think this just needs to be done earlier..